### ISSUE TITLE
Implement Learning Rate Optimizer

### ISSUE THREAD

**Issue by Cotidie** [Cotidie/ai-from-scratch#2]
# AS-IS
- learning rate is left as a pure hyper-parameter (defaults to 0.001)
- optimizer is needed to ensure faster convergence into local minima

# TO-BE
- [ ] Define an interface for Learning Rate Optimizer
- [ ] Implement AdaGrad optimizer
- [ ] Implement RMSProp optimizer
- [ ] Implement Moment optimizer
- [ ] Implement Adam optimizer
